make Alternative Title repeatable

lubitchv opened the pull request:
**What this PR does / why we need it**:
Make "Alternative Title" multiple (repeatable) field such that DDI export would not break.

**Which issue(s) this PR closes**:
Closes #9428 

**Special notes for your reviewer**:
If one does not want "Alternative Title" to be repeatable it is enough to remove TRUE from citation.tsv and solr, but leave the changes in the code that will allow for those who wants repeatable. 
**Suggestions on how to test this**:
Create a dataset with multiple alternative titles, publish it. Test DDI Exporter and OpenAire exporter. It should not break and should have multiple alternative titles. 

**Does this PR introduce a user interface change? If mockups are available, please link/include them here**:
Yes

**Is there a release notes update needed for this change?**:
Yes. 9428-alternative-title.md

lubitchv commented:
I do not have access to continuous integration testing server. Please let me know which integration tests are failing.

pdurbin commented:
@lubitchv yeah, it would be very nice to give this insight on the outside somehow. Here are some screenshots at least! From https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-9440/2/testReport/

<img width="1099" alt="Screenshot 2023-03-14 at 5 28 26 PM" src="https://user-images.githubusercontent.com/21006/225141028-5a878489-06aa-42e8-b618-88d6db769fe7.png">
<img width="1099" alt="Screenshot 2023-03-14 at 5 28 01 PM" src="https://user-images.githubusercontent.com/21006/225141036-901b40a4-e309-4259-9738-f93806a65e50.png">
<img width="1099" alt="Screenshot 2023-03-14 at 5 27 36 PM" src="https://user-images.githubusercontent.com/21006/225141039-14cf7940-5cc5-40ea-bec0-6d3f6bdb6c7e.png">

lubitchv commented:
The testImportDDI had error for productionPlace field (there was problem there with multiple). It was not related to this PR, but I fixed it. testArchivalStatusAPI and testLoadMetadataBlock_NoErrorPath I also fixed. But it seems that some integration tests are still failing  and I do not know which.

pdurbin commented:
@lubitchv this time ( https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-9440/5/ ) it wasn't API tests that failed. No API tests were run at all. I just clicked "re-run".

lubitchv commented:
API tests are now running fine.

pdurbin commented:
@lubitchv hi! Can you please resolve the merge conflicts in the PR?

There were probably caused by this PR being merged yesterday:

- #9484

lubitchv commented:
@pdurbin  I updated the branch from develop

lubitchv commented:
@pdurbin It says that there is 1 errored check, but I cannot see it, or I do not have permissions to see it.

pdurbin commented:
@lubitchv hi! Sorry for the noise. The failure in Jenkins is unrelated to your changes. It's this:

- https://github.com/GlobalDataverseCommunityConsortium/dataverse-ansible/issues/281

pdurbin commented on src/test/java/edu/harvard/iq/dataverse/export/ddi/dataset-create-new-all-ddi-fields.json:
```diff
@@ -51,9 +51,9 @@
+            "multiple": true,
             "typeClass": "primitive",
-            "value": "Alternative Title"
+            "value": ["Alternative Title"]
```
It looks like API users will encounter backward incompatibility. Is that right? If so, can this please be added to the release note snippet?

lubitchv replied:
Yes, repeatable altTitl is valid DDI. See https://ddialliance.org/Specification/DDI-Codebook/2.5/XMLSchema/field_level_documentation_files/schemas/codebook_xsd/elements/titlStmt.html#a19 It has maxOccurs="unbounded"

I updated dataset-finch xml and json and added multiple alternative titles such that validity test can test it.

I also added comment about incompatibility into release note.

landreev replied:
Could you please add a second value here, just like in dataset-finch1.json - 
```
["Alternative Title1", "Alternative Title2"]
```
and that will need a corresponding change in `src/test/java/edu/harvard/iq/dataverse/export/ddi/exportfull.xml` that DdiExportUtilTest compares against. 
(I'm putting together a list of all the changes that need to be made when fields are made repeatable, for other developers)
Thank you!

lubitchv replied:
@landreev I added ``["Alternative Title1", "Alternative Title2"]`` into dataset-create-new-all-ddi-fields.json and added another title in corresponding exportfull.xml

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/export/openaire/OpenAireExportUtil.java:
```diff
@@ -326,12 +322,36 @@ public static void writeTitlesElement(XMLStreamWriter xmlw, DatasetVersionDTO da
+    private static boolean writeMultipleTitleElement(XMLStreamWriter xmlw, String titleType, DatasetVersionDTO datasetVersionDTO, String metadataBlockName, boolean title_check, String language) throws XMLStreamException {
+        MetadataBlockDTO block = datasetVersionDTO.getMetadataBlocks().get(metadataBlockName);
+        if (block != null) {
+            logger.info("Block is not empty");
```
```suggestion
            logger.fine("Block is not empty");
```

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/export/openaire/OpenAireExportUtil.java:
```diff
@@ -326,12 +322,36 @@ public static void writeTitlesElement(XMLStreamWriter xmlw, DatasetVersionDTO da
+            List<FieldDTO> fieldsBlock =  block.getFields();
+            if (fieldsBlock != null) {
+                for (FieldDTO fieldDTO : fieldsBlock) {
+                    logger.info(titleType + " " + fieldDTO.getTypeName());
```
```suggestion
                    logger.fine(titleType + " " + fieldDTO.getTypeName());
```

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/export/openaire/OpenAireExportUtil.java:
```diff
@@ -326,12 +322,36 @@ public static void writeTitlesElement(XMLStreamWriter xmlw, DatasetVersionDTO da
+                for (FieldDTO fieldDTO : fieldsBlock) {
+                    logger.info(titleType + " " + fieldDTO.getTypeName());
+                    if (titleType.toLowerCase().equals(fieldDTO.getTypeName().toLowerCase())) {
+                        logger.info("Found Alt title");
```
```suggestion
                        logger.fine("Found Alt title");
```

pdurbin commented on src/main/java/edu/harvard/iq/dataverse/export/ddi/DdiExportUtil.java:
```diff
@@ -202,7 +202,11 @@ private static void createStdyDscr(XMLStreamWriter xmlw, DatasetDTO datasetDto)
         writeFullElement(xmlw, "titl", dto2Primitive(version, DatasetFieldConstant.title), datasetDto.getMetadataLanguage());
         writeFullElement(xmlw, "subTitl", dto2Primitive(version, DatasetFieldConstant.subTitle));
-        writeFullElement(xmlw, "altTitl", dto2Primitive(version, DatasetFieldConstant.alternativeTitle));
+        //writeFullElement(xmlw, "altTitl", dto2Primitive(version, DatasetFieldConstant.alternativeTitle));
```
```suggestion
```

pdurbin reviewed (commented):
Just some quick feedback and a few questions:

Is it valid in DDI to have more than one alternate title?

This pull request was recently merged and now Dataverse produces valid DDI (as far as I know):

- #9484

I see there's a test in that PR that calls `isValid`. Does that test cover changes in this PR? If not, can we please have a test that shows that the DDI is still valid?

jggautier commented:
@pdurbin Yeah, it's valid according to the schema rules at https://ddialliance.org/Specification/DDI-Codebook/2.5/XMLSchema/codebook.xsd. Here's a screenshot of the section in that XSD file where "maxOccurs" for altTitl is "unbounded":

<img width="667" alt="Screenshot 2023-05-15 at 12 39 03 PM" src="https://github.com/IQSS/dataverse/assets/18374574/7d601b4a-3399-4789-8f76-f5a80cb32c10">

pdurbin commented:
@siacus @cmbz @scolapasta can this be prioritized? @lubitchv attended the community call this morning and asked about it.

cmbz commented:
2023/08/02: During Kickoff, we decided to wait until 6.1 to address this.

landreev commented:
> The testImportDDI had error for productionPlace field (there was problem there with multiple).

Thank you for fixing this!

pdurbin commented:
@lubitchv when you get a chance can you please merge the latest from develop to pick up the upgrade to Java 17?  I posted about it here: https://groups.google.com/g/dataverse-dev/c/RnKAmpoGSf8/m/ebjAoAO8AAAJ

That should resolve the current failure in Jenkins: https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-9440/17/

Thanks!

lubitchv commented:
@pdurbin updated from develop, I see that there is some error, is it integration tests? Compilation on my local machine and unit tests are fine.

pdurbin commented:
@lubitchv thanks for merging develop.

I just kicked off https://github.com/gdcc/api-test-runner/actions/runs/5962550825

After it finishes, can you see the results?

pdurbin commented:
@poikilotherm when I open https://github.com/gdcc/api-test-runner/actions/runs/5962550825/job/16173949398 in a private window I can't see the details of the API test failure:

<img width="1582" alt="Screenshot 2023-08-24 at 6 53 10 AM" src="https://github.com/IQSS/dataverse/assets/21006/41ab6e73-1f08-4184-ba2a-e8afcaccee08">

Do you know if we can open this up somehow? Or maybe as a workaround for now we add @lubitchv to that test runner repo?

@lubitchv I don't mean to keep you in suspense. Here's the failure:

<img width="1582" alt="Screenshot 2023-08-24 at 6 55 41 AM" src="https://github.com/IQSS/dataverse/assets/21006/feabfa37-90bc-420b-b9a8-6934bec26c82">

@lubitchv I'm happy to help you get set up to run API tests on your laptop. Do you use Docker?

lubitchv commented:
MetadataBlocksIT seems to work now. I am not sure what does 1 cancelled and 2 skipped checks mean.

pdurbin commented:
@lubitchv yes! All tests passing now! Please don't worry about the skips and cancels. The skips are related to this:

- #9775

The cancel was me, sorry, related to this: https://github.com/IQSS/dataverse/pull/9830#issuecomment-1693550818

scolapasta commented:
@lubitchv now that we've released 6.0, I'm moving waiting 6.1 issues back to the board.

But could you first handle the 6.0 merge and address any EE10 issues. Thanks.

lubitchv commented:
Some integration tests are failing again. can I see which ones?

pdurbin commented:
@lubitchv the failing tests might be due to Jenkins/Ansible not being updated to 6.0 yet:

- https://github.com/GlobalDataverseCommunityConsortium/dataverse-ansible/issues/316

But I'm not sure. Here's the error I see at https://jenkins.dataverse.org/job/IQSS-Dataverse-Develop-PR/job/PR-9440/20/consoleFull

TASK [dataverse : fire off installer] ******************************************
[WARNING]: Module remote_tmp /home/dataverse/.ansible/tmp did not exist and was
created with a mode of 0700, this may cause issues when running as another
user. To avoid this, create the remote_tmp dir with the correct permissions
manually
fatal: [localhost]: FAILED! => {"changed": true, "cmd": "/usr/bin/python3 /tmp/dvinstall/install.py -f --config_file=default.config --noninteractive > /tmp/dvinstall/install.out 2>&1", "delta": "0:00:00.288717", "end": "2023-09-11 18:33:35.911959", "msg": "non-zero return code", "rc": 1, "start": "2023-09-11 18:33:35.623242", "stderr": "", "stderr_lines": [], "stdout": "", "stdout_lines": []}

Meanwhile, I tried a test run here: https://github.com/gdcc/api-test-runner/actions/runs/6150829101

pdurbin commented:
> Meanwhile, I tried a test run here: https://github.com/gdcc/api-test-runner/actions/runs/6150829101

All the tests passed: Tests run: 210, Failures: 0, Errors: 0, Skipped: 10

landreev reviewed (approved):
Looks perfect, thank you again!

kcondon commented:
I'm merging this, thanks! I did find a minor search results display bug relating to fields that have changed from single to multiple in metadata but will open a separate issue since it exists already. Thanks for the help @landreev

pdurbin commented:
Related:

- https://github.com/IQSS/dataverse-sample-data/pull/41
